Automated parameters & secrets

bananaoomarang opened the pull request:
Implements what was described [here](https://gist.github.com/bananaoomarang/76cf0e72911b38b4f9e0)

ggoodman commented on bin/create.js:
```diff
@@ -225,12 +280,12 @@ function handleCreate (argv) {
-                                + '/' + argv.name
-                                + (argv.prod ? '': '?webtask_no_cache=1');
+                                + '/' + tokenOpts.name
+                                + (argv.prod ? '': '&webtask_no_cache=1');
```
Need to move that `&` back to a `?`.

tjanczuk commented:
I like the idea of sourcing secrets from .env, this will greatly streamline webtask creation in dev mode. 

I am not so sure about rolling out the batch support right now. I think it would be good to see what usage patterns emerge to see if there is a real need. So I suggest to scope it to a single webtask for now. 

Also, @woloski had previously floated the idea of sourcing the parameter metadata from jsdoc embedded in the webtask code itself. I think this would make it much more natural than the "death by configuration" approach of package.json. Maybe you can explore what can be done with jsdoc?

bananaoomarang commented:
Yes, I think the package.json set up at the moment does feel overcomplicated. I will look into implementing a jsdoc solution now.

Re: batching, I was thinking of building a set of microservices, but you are right: I think in reality you would want to work on one task at once 90% of the time, since a separation of concerns is the aim there anyway.

bananaoomarang commented:
New version looks like this, and if no secrets config is declared everything from `.env` is passed. Look good?

```javascript
/**
 * @param  NOT_SECRET value
 * @param FOO bar
 * @secret MONGODB_URL
 * @secret PASSWORD
 */

module.exports = function (ctx, done) {
    done(null, ctx.data);
};
```

ggoodman commented:
That approach looks really sweet. This pulls from `process.env` and/or `.env`?

bananaoomarang commented:
Yup, it should do, can't test it at the moment because AWS is broken but will push later.

woloski commented:
:heart_eyes:

woloski commented:
We should be able to generate a webpage like this, with the jsdoc. Can you "run that test" @bananaoomarang?

![](https://camo.githubusercontent.com/e1fc5b0365e8edc87204c9d61720ef6893faf36d/68747470733a2f2f646f63732e676f6f676c652e636f6d2f64726177696e67732f642f31567948774561667946545169656d3064526d6f44666f395f52426e5f576f37556e6d5146445237577447302f7075623f773d39363026683d373230)

bananaoomarang commented:
That would be nice, though currently I'm using the `@param` description as a default value to pass in, though now it's in the webtask itself that doesn't make much sense.

tjanczuk commented:
Is the `@secret` an extension of jsdoc you invented? I wonder if we can keep things simple and only use `@param` to mean the secret parameters. No support for plaintext parameters at this level. 

Regarding defaults, I think people should be doing defaults in code directly not in jsdoc/configuration, e.g. 

```
context.data.foo = context.data.foo || 'my_default'
```

since this will cover the non-wt and non-jsdoc use cases as well. 

I think the `@param` description would be what the cli shows as a prompt, not the default value.

tjanczuk commented on bin/create.js:
```diff
@@ -154,7 +171,16 @@ function handleCreate (argv) {
+            if(fileOrDirectory.isFile()) {
+                argv.code = Fs.readFileSync(argv.file_name, 'utf8');
+            } else {
+                argv.code = walk(argv.file_name)
```
I thought you wanted to cut the directory support?

ggoodman replied:
I think that instead of blindly adding every `.js` file, we should give the user the ability to use glob patterns. This seems to be the way most such interfaces work. As @tjanczuk said, I would also prefer to see the directory adding in a separate PR.

tjanczuk commented on bin/create.js:
```diff
@@ -291,4 +352,104 @@ function parseHash (argv, field) {
+    var dotEnv;
+
+    try {
+        dotEnv = Fs.readFileSync('./.env', 'utf8');
```
I think it would be great to use the `dotenv` module to do this for you, and then use `process.env`. This will also have the side effect of allowing people to use env variables in addition to `.env` file to provide parameters.

bananaoomarang commented:
I like the idea that a page like what Matias posted could be generated from this, and losing support for plaintext `@param`s means they wouldn't be auto-documented. Since `ctx.data.param` is not transparent this could be confusing.

Re: default values, I agree on one level that it is probably more sensible to do it traditionally, but I also think it would be nice for developers if the jsdoc could be parsed to provide them, if it's already specifying secrets.

An alternative approach might be to document params based on the token, but then we lose the ability to have descriptions unless we embed that too. We also fragment param supplying and secret supplying in the dev workflow this way.

woloski commented:
I say it doesn't hurt having the support for multiple webtask files in a folder and it's already implemented. I've heard feedback from people who tried to do stuff with multiple webtasks that this would be useful (sandrino, milo, douce). Worst case, people won't use it.

bananaoomarang commented:
New jsdoc syntax is:

```
/** My Crazy Webtask
* @param MY_NOT_SECRET [default value to pass] Configures such and such
* @secret MY_SECRET
* etc...
*/

module.exports...
```

Where secrets are pulled by name from `.env`/`process.env`.

Passing a directory to `wt create` also still works, as well as watching all the files in it. The one caveat is not having two webtasks of the same file name, but an error is thrown and doesn't seem showstopping to me.

ggoodman commented on bin/create.js:
```diff
@@ -324,3 +403,139 @@ function parseHash (argv, field) {
+        });
+
+    if(key) {
+        if(!value) throw noKeyErr;
```
I think that it is plausible that someone might want to set a secret that is an empty string. Perhaps test `typeof value !== 'undefined'`.

ggoodman commented on bin/create.js:
```diff
@@ -366,3 +393,139 @@ function parseHash (argv, field) {
+        });
+
+    if(key) {
+        if(typeof value === 'undefined') throw noKeyErr;
```
Why pre-generate the error and only throw it here? Wouldn't the error's `stack` (source) be misleading in this case.

bananaoomarang commented:
Rebased to master and squashed most of the commits.

This alters `wt create` to support the following:

Standard JSDoc parsing, eg:

```
/**
 * Test task
 * @param {secret} MONGO_URL - Mongo database
 * @param {string} COLLECTION - Collection to read
 * @param {string} [optional=defaultValue] - Optional parameter
 */
```

* `{secret}`s are pulled from `process.env` and `.env`
* if a secret is not present the user is prompted

* `{string}`s are standard params.
* If required the user is prompted 
* if optional the default value is used
* If optional and no default value is provided they are left undefined

This means that a user can write a JSDoc in place of irritating `if(!ctx.data.foo)` blocks if they only want to know something was passed.

It's presence is also useful for [scaffolding](https://github.com/auth0/wt-cli/issues/5) as we can parse the same JSDoc there to provide a nice setup UX, and for auto-generating docs on the fly as in @woloski's example above.

ggoodman commented on bin/utils/getTaskConfig.js:
```diff
@@ -0,0 +1,131 @@
+                });
+        }
+    }, { concurrency: 1 })
+    .then(function () {
```
With `bluebird` you can do `.return(obj);`

ggoodman commented on bin/utils/getTaskConfig.js:
```diff
@@ -0,0 +1,131 @@
+function promptFor (type, obj) {
+    return Bluebird.map(Object.keys(obj), function (key) {
+        if(obj[key] === null) {
+            return Promptly.promptAsync('Please supply ' + type + ' ' + key.green + ':')
```
Please use `Promptly.passwordAsync()` when `type === 'secret'`.

bananaoomarang commented:
https://github.com/auth0/wt-cli/pull/25
